### DISCUSSION TITLE
Confused about fingerprints from signature factory

### DISCUSSION THREAD

**Discussion by jepdavidson** [rdkit/rdkit#5549] [Q&A]
Dear All,

I am currently trying to use the *feature factory* and *signature factory* mechanisms in RDKit to generate 2D pharmacophore fingerprints.  I thought I was there, but have realised that the fingerprints I am generating do not have any bits set for 2-point features...  Here's an example:

```python
from rdkit import Chem
from rdkit.Chem import ChemicalFeatures
from rdkit.Chem.Pharm2D import Generate
from rdkit.Chem.Pharm2D.SigFactory import SigFactory
from rdkit.RDConfig import RDDataDir


fdef_file=f"{RDDataDir}/BaseFeatures.fdef"
feat_factory = ChemicalFeatures.BuildFeatureFactory(fdef_file)
sig_factory = SigFactory(feat_factory, minPointCount=2, maxPointCount=3, useCounts=True)
sig_factory.SetBins([(0, 1), (1, 2), (2, 3), (3, 4), (4, 5), (5, 11)])
sig_factory.Init()

mol = Chem.MolFromSmiles("c1ccccn1")

fp = Generate.Gen2DFingerprint(mol, sig_factory)
print(f"Set bits:\t{fp.GetNonzeroElements()}")

families = sig_factory.GetFeatFamilies()
mol_feats = sig_factory.GetMolFeats(mol)
for feature, family in zip(mol_feats, families):
    if feature:
        print(f"{family}:\t{feature}")
```

> Set bits:	{}
> Acceptor:	[(5,)]
> Aromatic:	[(0, 1, 2, 3, 4, 5)]

With the example of pyridine, I was expecting the signature factory to generate a fingerprint with a bit set for the (0, 1) distance bin for Acceptor - Aromatic.  But no bits are set(?).

Indeed, with a little bit of (alphabetical) searching we can find that the signature factory knows **bit 6** represents this 2-point pharmacophore.

```python
print(sig_factory.GetBitDescription(6))
```

> Acceptor Aromatic |0 0|0 0|

So, my question is (as usual): what am I missing / doing wrong?!

And one other related question: is it reasonable to think a pharmacophoric fingerprint could have bits set for the individual features (eg a bit for Acceptor, a bit for Aromatic, etc)?  Of course, the distance bins would be irrelevant for these bins...  I thought it might work by setting `minPointCount=0` when constructing the signature factory - but that raises an error!

(The - somewhat niche - use case for the above would be to allow eg propane (`CCC`) and propanol (`CCCO`) to have a bit in common)

Kind regards

James

**Comment by greglandrum** [comment]
@jepdavidson what's happening here is that the 2D pharmacophore code does not allow zero distances. This is a quick way to remove overlapping or redundant pharmacophores.
We could theoretically add an option to remove this restriction, but it may have some additional knock-on effects.
Is this something you'd consider important ?

**Reply by jepdavidson** [reply]
Thanks @greglandrum - I probably should have figured that out(!).

> We could theoretically add an option to remove this restriction, but it may have some additional knock-on effects.
> Is this something you'd consider important ?

I guess the answer is "I *think* so...".  But now I am curious how other pharmacophoric fingerprints handle this situation(?).

In 3D (or even 2D), the 'shared' features would geometrically be > zero distance apart (with one representing the basic nitrogen, and one representing a projected point in the centre of the aromatic ring).  I certainly don't want to ignore/lose such feature relationships.

Indole ("N1C=CC2=C1C=CC=C2") is another reasonable example - I would view this as an ARYL-ARYL-DONOR triangle.  But with the zero-distance restriction I suppose the two aryl rings don't see each other, and the donor and 5-membered aryl don't see each other...

I would welcome the ability to turn-off the zero-distance restriction, but would also welcome someone that has thought about this for longer to tell me why this might not be a good idea! :)

Kind regards

James
